torch_xla/experimental/scan.py

```diff
@@ -82,14 +132,514 @@ def scan(fn, init, xs):
   if xs_length is None:
     raise ValueError(f"`xs` {xs} is an empty PyTree.")
 
-  carry = init
-  ys = []
+  forward, backward = value_and_grad_partitioned(
+      fn, init, xs, partition_fn=partition_fn)
+  carry, ys = Scan.apply(forward, backward, init, xs)  # type: ignore
+  return carry, ys
+
+
+def value_and_grad_partitioned(
+    fn: Callable[[Carry, X], tuple[Carry, Y]],
+    init: Carry,
+    xs: X,
+    partition_fn=default_partition) -> tuple[Callable, Callable]:
+  """
+  Given a user `fn` to be scanned over the major axis of the input `xs` PyTree
+  and an initial carry object `init`, symbolically traces `fn` and returns two
+  functions, `forward` and `backward`, where `forward` packs intermediate
+  activations in the output `y`, and `backward` reads intermediate activations
+  from the input `x`. `backward` also reads `grad_y` and outputs `grad_x` via
+  its output `y`. Finally, `backward` takes in `grad_carry` from its input `carry`
+  and returns new tangents through its output `carry`. More formally:
+
+    forward(carry, x) -> (new_carry, (y, activations))
+    backward(grad_new_carry, (grad_y, activations)) -> (grad_carry, grad_x)
+
+  where `grad_y` is the gradient w.r.t `y` from `forward`, and `grad_new_carry`
+  is the gradient w.r.t. `new_carry` from the `forward`.
+
+  `activations` will always be a flat list of tensors.
+
+  This is similar to the `value_and_grad` transform found in JAX, but additionally
+  partitions and returns separate forward/backward passes, so that we may later
+  use them in the `autograd.Function` implementation of `Scan`.
+  """
+
+  # Make some fake tensors to trace the user function and obtain the
+  # forward and backward graphs. Note that the init/carry fake tensor
+  # always requires grad. That's because even if the user passed in some
+  # `init` that does not require grad, we still want gradients to flow
+  # through the `carry` from one iteration of the user function to the
+  # next. In summary, the `carry` argument used to trace a user function
+  # to get a correct backward pass always requires grad.
+  def make_fake_tensor(v: torch.Tensor, requires_grad=True) -> torch.Tensor:
+    return torch.empty_like(
+        v, dtype=v.dtype, device=v.device, requires_grad=requires_grad)
+
+  fake_carry_pytree = tree_map(make_fake_tensor, init)
+  fake_x_pytree = tree_map(
+      lambda v: make_fake_tensor(v[0], requires_grad=v.requires_grad), xs)
+
+  with torch.enable_grad():
+    fw_compiler, get_fwd = _make_get_graph_compiler()
+    bw_compiler, get_bwd = _make_get_graph_compiler()
+    fn_compiled = aot_function(
```

Comment:
Will this fail if there are tensors within `fn` that were not provided as parameters, since they are not 'fake' tensors? Say, for model parameters if we have a module inside `fn` that we wish to also trace fwd/bwd on. Is this targeted as a follow-up?

Reply:
Yes, it will. That's why I added `scan_layers` (previously named `apply_layers`) in this PR to extract module parameters and functionalize the module.